Cutting transit graphs according to mwm.

bykoianko opened the pull request:
Данный PR обеспечивает при сборке mwm функциональность по перебору всех .transit.json с графами общественного транспорта, вырезанию их по границе обрабатываемой mwm согласно правилам из https://jira.mail.ru/browse/MAPSME-5858, и сохранению в mwm секцию.

PR уже прошу смотреть, однако в нем пока кое что не реализовано:

1. Сами правила по вырезанию графов обществного транспорта в методе GraphData::ClipGraphByMwm();
2. Использование visitors для ряда методов в GraphData;
3. Тесты на методы класса GraphData и на метод DeserializeFromJson;

Данные доработки будут добавлены в данный PR отдельными коммитами.

@ygorshenin @tatiana-kondakova PTAL

tatiana-kondakova commented:
>shapes: попадают только те shape, stop1_id и stop2_id будут в stops выше

если я правильно поняла, может так получиться, что некоторые shape никуда не попадут, не будут сохранены и нельзя будет передать их в drape

tatiana-kondakova commented:
Вычислительно эффективнее было бы один раз каждый json нарезать на mwm-ки, а потом при генерации соответсвующих mwm-ок использовать. Сейчас получается что если у нас есть большая сеть транспорта (RailEurope например), то вместо того чтобы обрабытывать её за `O(#стопов)` мы будем обрабатывать её за `O(#стопов * #mwm)` причем это даже не количество mwm покрытых сетью, а вообще мы этот json считаем и загрузим для каждой mwm-ки в мире, а потом уже в ClipGraphByMwm получим пустой граф. То есть в 1000 раз медленнее примерно. Это всё конечно на этапе генерации, но думаю всё равно получится заметное замедление.

ygorshenin commented on generator/transit_generator.cpp:
```diff
@@ -229,52 +158,20 @@ void DeserializerFromJson::operator()(StopIdRanges & rs, char const * name)
-  vector<Gate> gates;
-  DeserializeGatesFromJson(root, mwmDir, countryId, mapping, gates);
-  CHECK(IsValid(gates), (gates));
+  DeserializeItemFromJson(root, "stops", mapping, m_stops);
```
Is it possible to apply Visitor-s approach to GraphData too?

bykoianko replied:
Yes. It's already done.

ygorshenin commented on generator/transit_generator.cpp:
```diff
@@ -319,7 +208,179 @@ void BuildTransit(string const & mwmDir, string const & countryId,
+
+void GraphData::Append(GraphData const & rhs)
+{
+  m_stops.insert(m_stops.begin(), rhs.m_stops.begin(), rhs.m_stops.end());
```
Looks like you're doing not Append, but Prepend. Moreover, could you please implement following helper:
```c++
template <typename T>
void Append(vector<T> & src, vector<T> const & dst)
{
  src.insert(src.end(), dst.begin(), dst.end());
}
```
and use it? My eyes are in pain from all these inserts, begins and ends.

bykoianko replied:
Done, unfortunately I cannot use visitor here because of its param.

ygorshenin commented on generator/transit_generator.hpp:
```diff
@@ -102,11 +103,66 @@ class DeserializerFromJson
+class GraphData
+{
+public:
+  // @todo(bykoianko) All the methods below should be rewrite with the help of visitors.
```
rewritten?

bykoianko replied:
Yes. The todo removed since the visitors are written.

ygorshenin commented on generator/transit_generator.hpp:
```diff
@@ -102,11 +103,66 @@ class DeserializerFromJson
+class GraphData
+{
+public:
+  // @todo(bykoianko) All the methods below should be rewrite with the help of visitors.
```
Why not to to that right now?

bykoianko replied:
Done

ygorshenin commented on generator/transit_generator.hpp:
```diff
@@ -102,11 +103,66 @@ class DeserializerFromJson
+
+/// \brief Fills |data| according to a transit graph (|transitJsonPath|).
+/// \note Some of fields of |data| contains feature ids of a certain mwm. These fields are filled
+/// iff the mapping (|osmIdToFeatureIdsPath|) contains them. If not the fields have default value.
```
s/If not/Otherwise/

bykoianko replied:
Fixed.

ygorshenin commented on generator/transit_generator.hpp:
```diff
@@ -102,11 +103,66 @@ class DeserializerFromJson
+  /// if it's not valid.
+  /// \note |m_stops|, Edge::m_stop1Id and Edge::m_stop2Id in |m_edges| should be valid before call.
+  void CalculateEdgeWeight();
+  /// \brief Calculates best pedestrian segment for every gate in |m_gates|.
```
CalculateBestPedestrianSegments?

bykoianko replied:
Done

ygorshenin commented on generator/transit_generator.hpp:
```diff
@@ -102,11 +103,66 @@ class DeserializerFromJson
+  /// \brief Calculates and updates |m_edges| by adding valid value for Edge::m_weight
+  /// if it's not valid.
+  /// \note |m_stops|, Edge::m_stop1Id and Edge::m_stop2Id in |m_edges| should be valid before call.
+  void CalculateEdgeWeight();
```
CalculateEdgeWeights?

bykoianko replied:
Done

ygorshenin commented on generator/transit_generator.hpp:
```diff
@@ -102,11 +103,66 @@ class DeserializerFromJson
+  void CalculateBestPedestrianSegment(string const & mwmPath, string const & countryId);
+  /// \brief Removes some items from all the class fields if they outside |mwmBorders|.
+  /// @todo(bykoinko) Certain rules which is used to clip the transit graph should be described here.
+  void ClipGraphByMwm(std::vector<m2::RegionD> const & mwmBorders);
```
ClipGraph?

ygorshenin replied:
s/mwmBorders/borders/
Am I right that your code somehow depends on the fact that mwmBorders are real borders for some mwm? Or this code works with any borders? If so, please, remove "mwm" prefix.

bykoianko replied:
This method was written not to depend from the fact if the border is real mwm borders on not. 

But after @tatiana-kondakova comments I decided to try to move the clipping step to generate features step of generator. In that case all transit graph mwm would be processed only once for build of planet. 

The idea is to take all transit jsons on the generate features step, to split them into json according to mwm borders, and to create transit and transit cross mwm sections on this stage.

What do you think about that?

ygorshenin commented on generator/transit_generator.cpp:
```diff
@@ -319,7 +208,179 @@ void BuildTransit(string const & mwmDir, string const & countryId,
+    }
+    catch (RootException const & e)
+    {
+      LOG(LDEBUG, ("Point of a gate belongs to several mwms or doesn't belong to any mwm. Gate:",
```
LERROR please.

ygorshenin replied:
Also, how about to return false or re-throw the exception?

ygorshenin replied:
Pokemon exception handling? https://blog.codinghorror.com/new-programming-jargon/

How you may be sure that RootExceptions means only that "Point of a gate belongs to several mwms or ..."?

bykoianko replied:
The exception is raised by the following code:
```c++
bool IndexRouter::FindBestSegment(...l) const
{
  auto const file = platform::CountryFile(m_countryFileFn(point));
  MwmSet::MwmHandle handle = m_index.GetMwmHandleByCountryFile(file);
  if (!handle.IsAlive())
    MYTHROW(MwmIsNotAlifeException, ("Can't get mwm handle for", file));
...
```
So, |file| may be set to CountryFile("") if according to map data gates are located outside of simplified(!) mwm border. I write LOG(LERROR, ...) but are you sure that it's enough for LOG(LERROR, ...)?

bykoianko replied:
> Also, how about to return false or re-throw the exception?

It's a possible situation that the best segment is not found and it depends on the map. In that case I prefer keep building mwms.

ygorshenin replied:
My main point is that error message may be wrong and misleading. Are you sure that this is the only place where exception is thrown now? If yes, how you can guarantee that in future there will be the only place where exception is thrown? And why do you catch RootException instead of RoutingException?

bykoianko replied:
> And why do you catch RootException instead of RoutingException? 

Please have a look at the code. I catch a special MwmIsNotAlifeException derived from RoutingException now. Besides that I rewrote the error message.

bykoianko replied:
In PR which is going after this one I check if points of gates lay within processed mwm before to call FindBestSegmentInSingleMwm(...) which may raise an exception. I wrote:
```c++
...
  for (auto & gate : m_gates)
  {
    if (countryFileGetter(gate.GetPoint()) != countryId)
      continue;
    try
    {
      if (indexRouter.FindBestSegmentInSingleMwm(..))
      {
        CHECK_EQUAL(bestSegment.GetMwmId(), 0, ());
        gate.SetBestPedestrianSegment(SingleMwmSegment(
            bestSegment.GetFeatureId(), bestSegment.GetSegmentIdx(), bestSegment.IsForward()));
      }
    }
    catch (MwmIsNotAliveException const & e)
    {
      LOG(LCRITICAL, ("Point of a gate belongs to the processed mwm:", countryId, ","
          "but the mwm is not alive. Gate:", gate, e.what()));
    }
...
```

bykoianko replied:
I did it in this PR

tatiana-kondakova commented on generator/transit_generator.cpp:
```diff
@@ -174,6 +90,18 @@ void FillOsmIdToFeatureIdsMap(string const & osmIdToFeatureIdsPath, OsmIdToFeatu
+  return my::JoinFoldersToPath(mwmDir, countryId + DATA_FILE_EXTENSION);
+}
+
+void LoadBorders(string const & dir, string const countryId, vector<m2::RegionD> & borders)
```
`string const & countryId`

bykoianko replied:
Done

tatiana-kondakova commented on generator/transit_generator.hpp:
```diff
@@ -102,11 +103,66 @@ class DeserializerFromJson
+  void CalculateEdgeWeight();
+  /// \brief Calculates best pedestrian segment for every gate in |m_gates|.
+  /// \note All gates in |m_gates| should have a valid |m_point| field before the call.
+  void CalculateBestPedestrianSegment(string const & mwmPath, string const & countryId);
```
`std::string`

bykoianko replied:
Done

tatiana-kondakova commented on generator/transit_generator.hpp:
```diff
@@ -102,11 +103,66 @@ class DeserializerFromJson
+/// \brief Fills |data| according to a transit graph (|transitJsonPath|).
+/// \note Some of fields of |data| contains feature ids of a certain mwm. These fields are filled
+/// iff the mapping (|osmIdToFeatureIdsPath|) contains them. If not the fields have default value.
+void DeserializeFromJson(OsmIdToFeatureIdsMap const & mapping, string const & transitJsonPath,
```
`std::string`

bykoianko replied:
Fixed.

tatiana-kondakova commented on generator/transit_generator.hpp:
```diff
@@ -102,11 +103,66 @@ class DeserializerFromJson
+
+/// \brief Calculates and adds some information to transit graph (|data|) after deserializing
+/// from json.
+void ProcessGraph(string const & mwmPath, string const & countryId,
```
`std::string`

bykoianko replied:
Done

bykoianko commented on generator/transit_generator.cpp:
```diff
@@ -319,7 +208,179 @@ void BuildTransit(string const & mwmDir, string const & countryId,
+void ProcessGraph(string const & mwmPath, string const & countryId,
+                  OsmIdToFeatureIdsMap const & osmIdToFeatureIdsMap, GraphData & data)
+{
+  data.CalculateBestPedestrianSegment(mwmPath, countryId);
```
Я попробую вынести разбиение всех transit json-ов по mwm-кам на generate feature step. Из данного PR функциональность по разбиению графов общественного транспорта по mwm будет удалена.

bykoianko commented:
Как верно отметила @tatiana-kondakova имеет смысл разбивать графы общественного транспорта по mwm-кам один раз за все сборку планеты, а не для каждой mwm-ки.

Я попробую сделать это в следующем в следующем PR на feature generation шаге генератора. Результаты разбиения я сохраню в json-ах, в формате, аналогичном входному формату графа общественного транспорта в генератор. Поскольку все id глобально уникальны, наложения ряда дорожных сетей (графов) общественного транспорта в одну пройдет без коллизий.

С данным PR я ответил на все комменты.

@tatiana-kondakova @ygorshenin PTAL

ygorshenin commented on generator/transit_generator.cpp:
```diff
@@ -49,121 +48,92 @@ using namespace std;
+  {
+    sort(c.begin(), c.end());
+  }
+  void operator()(vector<transit::Edge> & c, char const * /* name */ = nullptr) const {}
```
Blank line, please.

bykoianko replied:
Fixed.

ygorshenin commented on generator/transit_generator.cpp:
```diff
@@ -49,121 +48,92 @@ using namespace std;
+  template <typename Cont>
+  void operator()(Cont const & c, char const * /* name */ = nullptr)
+  {
+    if (!m_isValid)
```
m_isValid = m_isValid && ::IsValid(c);

bykoianko replied:
Done

ygorshenin commented on generator/transit_generator.cpp:
```diff
@@ -49,121 +48,92 @@ using namespace std;
+{
+  ASSERT(is_sorted(stops.cbegin(), stops.cend()), ());
+  auto s1Id = equal_range(stops.cbegin(), stops.cend(),
+          Stop(stopId, kInvalidOsmId, kInvalidFeatureId, kInvalidTransferId, {}, m2::PointD(), {}));
```
Could you please replace ```m2::PointD()``` by ```{}```?

bykoianko replied:
Done

ygorshenin commented on generator/transit_generator.cpp:
```diff
@@ -319,7 +256,182 @@ void BuildTransit(string const & mwmDir, string const & countryId,
+
+void GraphData::Clear()
+{
+  ClearVisitor const v;
```
```ClearVisitor{}(v);``` ?

bykoianko replied:
ClearVisitor doesn't have operator()(GraphData &). It only can call  method clear() for fields.

ygorshenin replied:
OK

ygorshenin commented on generator/transit_generator.cpp:
```diff
@@ -49,121 +48,92 @@ using namespace std;
+    if (!m_isUnique)
+      return;
+
+    m_isUnique = (adjacent_find(c.begin(), c.end()) == c.end());
```
Will be it useful to, at least, emit warn messages for containers that contain duplicates? In this case, it may be useful to get rid of early exit.

bykoianko replied:
A duplicate mean a significant error which is enough to stop generator. So if |m_isUnique| == false GraphData::IsValid() returns false and generator will be terminate in ProcessGraph() (called from BuildTransit())

Probably it's worth renaming methods of GraphData to IsValid(), IsUnique() and IsSorted() to CheckValid(), CheckUnique() and CheckSorted() and place CHECK inside them. 

What do you think about it?

ygorshenin commented on routing/index_router.cpp:
```diff
@@ -692,7 +693,7 @@ bool IndexRouter::FindBestSegment(m2::PointD const & point, m2::PointD const & d
   MwmSet::MwmHandle handle = m_index.GetMwmHandleByCountryFile(file);
   if (!handle.IsAlive())
-    MYTHROW(RoutingException, ("Can't get mwm handle for", file));
+    MYTHROW(MwmIsNotAlifeException, ("Can't get mwm handle for", file));
```
Alife?

bykoianko replied:
Alive

tatiana-kondakova commented on generator/transit_generator.cpp:
```diff
@@ -49,121 +48,92 @@ using namespace std;
+  {
+    sort(c.begin(), c.end());
+  }
+  void operator()(vector<transit::Edge> & c, char const * /* name */ = nullptr) const {}
```
У transit::Edge есть набор из трёх полей, которые уникально его идентифицируют (stop1Id, stop2Id, lineId), договорились Edge тоже сортировать и проверять на уникальность на этапе генерации -- это позволит на раннем этапе находить ошибки в данных и избавит от не вполне интуитивного поведения SortVisitor и IsSortVisitor

bykoianko replied:
Готово. Теперь у нас дуга будет однозначно идентифицироваться stop1 id, stop2 id и line id. Я обсудил этот вопрос со @Zverik и мы решили, что будем завершать генератор, если это не так.

mpimenov commented on generator/transit_generator.cpp:
```diff
@@ -213,12 +194,13 @@ void DeserializerFromJson::operator()(FeatureIdentifiers & id, char const * name
+  {
+    CHECK_EQUAL(it->second.size(), 1,
+                ("Osm id:", osmId, "(encoded", osmId.EncodedId(),
+                 ") from transit graph doesn't present by a single feature in mwm."));
```
s/doesn't present/isn't represented/

mpimenov replied:
Though, I don't understand what you mean by this.
Should it be represented by exactly one feature? If yes, don't we need two separate messages for the
```< 1``` and ```> 1``` cases?
Or,
```... must be represented by exactly one feature in the mwm.```

bykoianko replied:
As you know some osm features (osm id) may be converted to several our features (feature ids). For example: http://www.openstreetmap.org/way/42798631#map=18/56.29197/43.86209

Every osm feature from transit graph json should be converted to only one our feature. But in case of features near mwm border it's possible that there's no feature id for osm id.

In short, every osm id should be converted to 0 or 1 feature id.

mpimenov replied:
Then,
```is not represented by any feature in the mwm.```

mpimenov commented on generator/transit_generator.cpp:
```diff
@@ -319,7 +256,182 @@ void BuildTransit(string const & mwmDir, string const & countryId,
   LOG(LINFO, (TRANSIT_FILE_TAG, "section is ready. Header:", header));
 }
+
+void GraphData::Append(GraphData const & rhs)
```
s/Append/AppendTo/

I would expect that 
* ```a.Append(b)``` appends ```b``` to ```a```.
* ```a.AppendTo(b)``` appends ```a``` to ```b```.

bykoianko replied:
Done

mpimenov commented on generator/transit_generator.hpp:
```diff
@@ -102,15 +104,74 @@ class DeserializerFromJson
 /// \param osmIdToFeatureIdsPath is a path to a file with osm id to feature ids mapping.
-/// \param transitDir a path to directory with json files with transit graphs.
+/// \param transitDir a path with slash at the end to directory with json files with transit graphs.
+/// It's assumed that the files have name the same with country ids and extension TRANSIT_FILE_EXTENSION.
```
s/name the same/the same name/

bykoianko replied:
Done

mpimenov commented on routing_common/transit_types.cpp:
```diff
@@ -161,6 +161,13 @@ bool Gate::IsValid() const
+bool ShapeId::operator<(ShapeId const & rhs) const
+{
+  if (m_stop1Id != rhs.m_stop1Id)
+    return rhs.m_stop1Id < m_stop1Id;
```
Is it important that the order is reversed for stop1?

bykoianko replied:
It's a mistake. Fixed. Thanks.

bykoianko commented on generator/transit_generator.cpp:
```diff
@@ -49,121 +48,92 @@ using namespace std;
-  return *s1Id.first;
-}
+  template <typename Cont>
+  void operator()(Cont & c, char const * /* name */ = nullptr) const { c.clear(); }
```
I removed all the default initialization from all the visitors.

bykoianko commented on generator/transit_generator.cpp:
```diff
@@ -174,6 +144,17 @@ void FillOsmIdToFeatureIdsMap(string const & osmIdToFeatureIdsPath, OsmIdToFeatu
+
+string GetMwmPath(string const & mwmDir, string const & countryId)
+{
+  return my::JoinFoldersToPath(mwmDir, countryId + DATA_FILE_EXTENSION);
```
Done. In two places.

bykoianko commented:
@ygorshenin @tatiana-kondakova @mpimenov  PTAL

bykoianko commented:
Rebased. @ygorshenin @mpimenov PTAL

mpimenov commented on generator/transit_generator.cpp:
```diff
@@ -4,18 +4,17 @@
 
 #include "traffic/traffic_cache.hpp"
 
+#include "routing/routing_exceptions.hpp"
```
sort, please

bykoianko replied:
Done

mpimenov commented on generator/transit_generator.cpp:
```diff
@@ -49,121 +48,79 @@ using namespace std;
+  auto s1Id = equal_range(stops.cbegin(), stops.cend(),
+                          Stop(stopId, kInvalidOsmId, kInvalidFeatureId, kInvalidTransferId,
+                               {} /* line ids */, {} /* point */, {} /* title anchors */));
+  CHECK(s1Id.first != stops.cend(), ("No a stop with id:", stopId, "in stops:", stops));
```
s/a stop/stop/

bykoianko replied:
Done

mpimenov commented on generator/transit_generator.cpp:
```diff
@@ -321,7 +245,182 @@ void BuildTransit(string const & mwmDir, string const & countryId,
+  {
+    GetPlatform().GetReader(transitJsonPath)->ReadAsString(jsonBuffer);
+  }
+  catch (RootException const & ex)
```
Why RootException here?

bykoianko replied:
It's for case if a file is absent. As far as I'm aware GetReader() may raise `FileAbsentException`. But I'm not sure about others. 

It's a generator code. Let's catch only `FileAbsentException`. If we miss some exceptions the generator will terminate and we know about that.

mpimenov commented on generator/transit_generator.hpp:
```diff
@@ -102,15 +104,74 @@ class DeserializerFromJson
+  void Clear();
+  bool IsValid() const;
+
+  /// \brief Sorts all class fields execpt for |m_edges| by their ids.
```
s/execpt/except/

bykoianko replied:
Fixed.

mpimenov commented on generator/transit_generator.hpp:
```diff
@@ -102,15 +104,74 @@ class DeserializerFromJson
+  void Sort();
+  /// \brief Calculates and updates |m_edges| by adding valid value for Edge::m_weight
+  /// if it's not valid.
+  /// \note |m_stops|, Edge::m_stop1Id and Edge::m_stop2Id in |m_edges| should be valid before call.
```
s/should/must/ ?
Are you guessing or is it a requirement?

bykoianko replied:
must. It's a requirement.

mpimenov commented on generator/transit_generator.hpp:
```diff
@@ -102,15 +104,74 @@ class DeserializerFromJson
+  /// \note |m_stops|, Edge::m_stop1Id and Edge::m_stop2Id in |m_edges| should be valid before call.
+  void CalculateEdgeWeights();
+  /// \brief Calculates best pedestrian segment for every gate in |m_gates|.
+  /// \note All gates in |m_gates| should have a valid |m_point| field before the call.
```
s/should/must/ ?

bykoianko replied:
must

mpimenov commented on generator/transit_generator.hpp:
```diff
@@ -102,15 +104,74 @@ class DeserializerFromJson
+  std::vector<Network> const & GetNetworks() const { return m_networks; }
+
+private:
+  DECLARE_VISITOR_AND_DEBUG_PRINT(GraphData, visitor(m_stops, "stops"), visitor(m_gates, "gates"),
```
Consider overriding clang-format here leaving one visitor per line.

bykoianko replied:
Yes it's more readable but less convenient to edit. (I'm using special plugin in my IDE for clang-format)...

Probably I'll do it later in a separate PR and for all transit visitors.

mpimenov commented on generator/transit_generator.hpp:
```diff
@@ -102,15 +104,74 @@ class DeserializerFromJson
+};
+
+/// \brief Fills |data| according to a transit graph (|transitJsonPath|).
+/// \note Some of fields of |data| contains feature ids of a certain mwm. These fields are filled
```
s/contains/contain/

bykoianko replied:
Fixed

mpimenov commented on generator/transit_generator.hpp:
```diff
@@ -102,15 +104,74 @@ class DeserializerFromJson
 /// \param countryId is an mwm name without extension of the processed mwm.
 /// \param osmIdToFeatureIdsPath is a path to a file with osm id to feature ids mapping.
-/// \param transitDir a path to directory with json files with transit graphs.
+/// \param transitDir a path with slash at the end to directory with json files with transit graphs.
```
Can we use my::AddSlashIfNeeded and drop this requirement?

bykoianko replied:
It's an old requirement. I removed it from the comment.

bykoianko commented:
@ygorshenin @mpimenov PTAL

ygorshenin commented on generator/transit_generator.cpp:
```diff
@@ -49,121 +48,79 @@ using namespace std;
+  template <typename Cont>
+  void operator()(Cont const & c, char const * /* name */)
+  {
+    m_isUnique = m_isUnique && (adjacent_find(c.begin(), c.end()) == c.end());
```
cbegin, cend?

bykoianko replied:
Yes

ygorshenin commented on generator/transit_generator.cpp:
```diff
@@ -49,121 +48,79 @@ using namespace std;
-      LOG(LDEBUG, ("Point of a gate belongs to several mwms or doesn't belong to any mwm. Gate:",
-                   gate, e.what()));
-    }
+    m_isSorted = m_isSorted && is_sorted(c.begin(), c.end());
```
cbegin, cend?

bykoianko replied:
Yes

ygorshenin commented on generator/transit_generator.cpp:
```diff
@@ -213,12 +181,13 @@ void DeserializerFromJson::operator()(FeatureIdentifiers & id, char const * name
+  {
+    CHECK_EQUAL(it->second.size(), 1,
+                ("Osm id:", osmId, "(encoded", osmId.EncodedId(),
+                 ") from transit graph is not represented by any feature in the mwm."));
```
What if second.size() is 2? Or 3? Is it possible? If so, the error message is not correct.

bykoianko replied:
I rewrote the error message:
```
    CHECK_EQUAL(it->second.size(), 1, ("Osm id:", osmId, "(encoded", osmId.EncodedId(),
                 ") from transit graph is correspond to several features or is not represented by "
                 "any feature in the mwm."));
```

ygorshenin replied:
Note that it would be helpful to log it->second too.

bykoianko replied:
I wrote this way:
```
    CHECK_EQUAL(it->second.size(), 1, ("Osm id:", osmId, "(encoded", osmId.EncodedId(),
                 ") from transit graph is correspond to", it->second.size()" features."
                 "But osm id should be represented be one feature."));
```

ygorshenin commented on generator/transit_generator.cpp:
```diff
@@ -321,7 +245,182 @@ void BuildTransit(string const & mwmDir, string const & countryId,
+    }
+    catch (MwmIsNotAliveException const & e)
+    {
+      LOG(LERROR, ("Point of a gate belongs doesn't belong to any mwm according to "
```
belongs doesn't belong?

bykoianko replied:
I rewrote the message and change the code.

ygorshenin commented on generator/transit_generator.cpp:
```diff
@@ -321,7 +245,182 @@ void BuildTransit(string const & mwmDir, string const & countryId,
+  {
+    GetPlatform().GetReader(transitJsonPath)->ReadAsString(jsonBuffer);
+  }
+  catch (FileAbsentException const & ex)
```
What if file exists but can't be read?

ygorshenin replied:
As you're going to fail here in any case, I suggest you to catch all Reader exceptions here.

bykoianko replied:
In that case I can catch Reader::Exception:
```
    catch (MwmIsNotAliveException const & e)
    {
      LOG(LCRITICAL, ("Point of a gate belongs to the processed mwm:", countryId, ","
          "but the mwm is not alive. Gate:", gate, e.what()));
    }
    catch (Reader::Exception const & e)
    {
      LOG(LCRITICAL, ("Reader exception. CountryId::", countryId, ". Gate:", gate, e.what()));
    }
```

bykoianko replied:
On the other hand why do we need handle them? In case of such exception we should stop the generator.

bykoianko replied:
Let's catch all (RootException) here and stop generator in this case.

ygorshenin commented on generator/transit_generator.cpp:
```diff
@@ -321,7 +245,182 @@ void BuildTransit(string const & mwmDir, string const & countryId,
+  CHECK(root.get() != nullptr, ("Cannot parse the json file:", transitJsonPath));
+
+  data.Clear();
+  data.DeserializeFromJson(root, mapping);
```
Who'll handle json exceptions?

bykoianko replied:
Nobody. In case of json issues the generator will be stoped.

bykoianko replied:
Let's catch deserializer exception here and and stop the generator at this point.

ygorshenin commented on routing/routing_exceptions.hpp:
```diff
@@ -6,4 +6,5 @@ namespace routing
 {
 DECLARE_EXCEPTION(CorruptedDataException, RootException);
 DECLARE_EXCEPTION(RoutingException, RootException);
+DECLARE_EXCEPTION(MwmIsNotAliveException, RootException);
```
How about to create a base class for all routing exceptions? I.e, you may do following:
```c++
DECLARE_EXCEPTION(RoutingException, RootException);
DECLARE_EXCEPTION(MwmIsNotAliveException, RoutingException);
DECLARE_EXCEPTION(CorruptedDataException, RoutingException);
```
?
What do you think?

bykoianko replied:
I think it's a good idea. Let's do it in this PR.

bykoianko replied:
I overlooked the code. Everything should be ok after this changes.

mpimenov commented on routing_common/transit_types.cpp:
```diff
@@ -165,6 +165,13 @@ bool Gate::IsValid() const
+bool ShapeId::operator<(ShapeId const & rhs) const
+{
+  if (m_stop1Id != rhs.m_stop1Id)
+    return m_stop1Id < rhs.m_stop1Id;
```
Let's also write a test for this.
Also, are there tests for Unique() and Valid()?

bykoianko replied:
I added test on IsValid() for transit data structure at transit_types.
Besides that I fixed some issues about IsValid() and add some more strict checks for IsValid() according to agreement with @Zverik PTAL.

As for Unique these tests will be write later, with a big tests on GraphData.

bykoianko replied:
> Let's also write a test for this.

Done

bykoianko commented:
I added test on IsValid() for transit data structure at transit_types.
Besides that I fixed some issues about IsValid() and add some more strict checks for IsValid() according to agreement with @Zverik PTAL.

bykoianko commented:
@ygorshenin @tatiana-kondakova @mpimenov PTAL

mpimenov commented on generator/transit_generator.cpp:
```diff
@@ -321,7 +245,196 @@ void BuildTransit(string const & mwmDir, string const & countryId,
+    }
+    catch (RootException const & e)
+    {
+      LOG(LCRITICAL, ("Exception while looking for the best segment here gates. CountryId::",
```
"best segment here gates"? What does it mean?

bykoianko replied:
`Exception while looking for the best segment of a gate.`

mpimenov commented on generator/transit_generator.cpp:
```diff
@@ -321,7 +245,196 @@ void BuildTransit(string const & mwmDir, string const & countryId,
+    }
+    catch (RootException const & e)
+    {
+      LOG(LCRITICAL, ("Exception while looking for the best segment here gates. CountryId::",
```
s/::/:/

bykoianko replied:
Fixed.

mpimenov commented on generator/transit_generator.hpp:
```diff
@@ -102,15 +104,74 @@ class DeserializerFromJson
+};
+
+/// \brief Fills |data| according to a transit graph (|transitJsonPath|).
+/// \note Some of fields of |data| contain feature ids of a certain mwm. These fields are filled
```
s/Some of/Some/

bykoianko replied:
Fixed.

mpimenov commented on routing_common/routing_common_tests/transit_test.cpp:
```diff
@@ -133,71 +162,94 @@ UNIT_TEST(Transit_SingleMwmSegmentSerialization)
+
+  TEST_LESS(id1, id2, ());
+  TEST_LESS(id2, id3, ());
+  TEST_LESS(id3, id4, ());
```
nit: make an array and test that it is sorted

bykoianko replied:
Done

bykoianko commented on generator/transit_generator.cpp:
```diff
@@ -213,12 +181,13 @@ void DeserializerFromJson::operator()(FeatureIdentifiers & id, char const * name
+  if (it != m_osmIdToFeatureIds.cend())
+  {
+    CHECK_EQUAL(it->second.size(), 1, ("Osm id:", osmId, "(encoded", osmId.EncodedId(),
+                 ") from transit graph is correspond to", it->second.size(), "features."
```
Fixed.

bykoianko commented:
@mpimenov PTAL

mpimenov commented on generator/transit_generator.cpp:
```diff
@@ -321,7 +245,196 @@ void BuildTransit(string const & mwmDir, string const & countryId,
+
+void GraphData::CalculateEdgeWeights()
+{
+  CHECK(is_sorted(m_stops.cbegin(), m_stops.cend()), ());
```
Isn't CHECK too heavy here?
Also, if this is going to be the only place where FindStopById is used (and it is now),
please remove ASSERT(is_sorted) from FindStopById.

bykoianko replied:
Yes. Let's remove it from here. Anyway it's checked in ProcessGraph() now. After CalculateEdgeWeights() calls.

bykoianko replied:
Done

bykoianko commented:
@mpimenov PTAL

bykoianko commented:
@mpimenov PTAL
